libcodechecker/log/log_parser.py

```diff
@@ -0,0 +1,791 @@
+    """
+    This function skips the flag pointed by the given flag_iterator with its
+    parameters if any.
+    """
+    if IGNORED_OPTIONS.match(flag_iterator.item):
+        return True
+
+    for pattern, arg_num in IGNORED_PARAM_OPTIONS.items():
+        if re.match(pattern, flag_iterator.item):
+            for i in range(arg_num):
+                next(flag_iterator)
+            return True
+
+    return False
+
+
+def parse_options(compilation_db_entry):
+    """
+    This function parses a GCC compilation action and returns a BuildAction
+    object which can be the input of Clang analyzer tools.
+
+    compilation_db_entry -- An entry from a valid compilation database JSON
+                            file, i.e. a dictionary with the compilation
+                            command, the compiled file and the current working
+                            directory.
+    """
+
+    buildaction = BuildAction()
+
+    if 'arguments' in compilation_db_entry:
+        gcc_command = compilation_db_entry['arguments']
+        buildaction.original_command = ' '.join(gcc_command)
+    elif 'command' in compilation_db_entry:
+        buildaction.original_command = compilation_db_entry['command']
+        gcc_command = compilation_db_entry['command'] \
+            .replace(r'\"', '"') \
+            .replace(r'"', r'"\"')
+        gcc_command = shlex.split(gcc_command)
+    else:
+        raise KeyError("No valid 'command' or 'arguments' entry found!")
+
+    buildaction.directory = compilation_db_entry['directory']
+    buildaction.action_type = buildaction.COMPILE
+    buildaction.compiler = gcc_command[0]
+    if '++' in buildaction.compiler or 'cpp' in buildaction.compiler:
+        buildaction.lang = 'c++'
+
+    flag_transformers = [
+        __get_arch,
+        __get_language,
+        __get_output,
+        __determine_action_type,
+        __replace,
+        __skip,
+        __append_from_file,
+        __collect_compile_opts,
+        __collect_sources]
+
+    for it in OptionIterator(gcc_command[1:]):
+        for flag_transformer in flag_transformers:
+            if flag_transformer(it, buildaction):
+                break
+        else:
+            pass
+            # print('Unhandled argument: ' + it.item)
+
+    def abs_path_maker(p):
+        return os.path.normpath(
+            os.path.join(compilation_db_entry['directory'], p))
+
+    merge_space_files(buildaction.sources, compilation_db_entry['file'])
+    buildaction.sources = map(abs_path_maker, buildaction.sources)
+
+    compiled_file = abs_path_maker(compilation_db_entry['file'])
+    if compiled_file != '.' and \
+            compiled_file not in map(abs_path_maker, buildaction.sources):
+        buildaction.sources.append(compiled_file)
+
+    for source_file in buildaction.sources:
+        lang = get_language(os.path.splitext(source_file)[1])
+        if lang:
+            if buildaction.lang is None:
+                buildaction.lang = lang
```

Comment:
If there are multiple files the language can be different for each file!
Do we have a separate entry in the compilation database for each file in this case? If so, we should rather use the compilation_db_entry['file'] to determine the language rather than the list of source. Moreover, if this is the case, we should not have a list of sources per build action but rather only have one single source.